spring-cloud-stream-binder-kafka-core/src/main/java/org/springframework/cloud/stream/binder/kafka/properties/KafkaBinderConfigurationProperties.java

```diff
@@ -132,9 +157,44 @@ public Transaction getTransaction() {
 	}
 
 	public String getKafkaConnectionString() {
+		// We need to do a check on certificate file locations to see if they are given as classpath resources.
+		// If that is the case, then we will move them to a file system location and use those as the certificate locations.
+		// This is due to a limitation in Kafka itself in which it doesn't allow reading certificate resources from the classpath.
+		// See this: https://issues.apache.org/jira/browse/KAFKA-7685
+		// and this: https://cwiki.apache.org/confluence/display/KAFKA/KIP-398%3A+Support+reading+trust+store+from+classpath
+		moveCertsToFileSystemIfNecessary();
+
 		return toConnectionString(this.brokers, this.defaultBrokerPort);
 	}
 
+	private void moveCertsToFileSystemIfNecessary() {
+		try {
+			final String trustStoreLocation = this.configuration.get("ssl.truststore.location");
+			if (trustStoreLocation != null && trustStoreLocation.startsWith("classpath:")) {
+				final String fileSystemLocation = moveCertToFileSystem(trustStoreLocation, this.truststoreLocationOnFileSystem);
+				// Overriding the value with absolute filesystem path.
+				this.configuration.put("ssl.truststore.location", fileSystemLocation);
+			}
+			final String keyStoreLocation = this.configuration.get("ssl.keystore.location");
+			if (keyStoreLocation != null && keyStoreLocation.startsWith("classpath:")) {
+				final String fileSystemLocation = moveCertToFileSystem(keyStoreLocation, this.keystoreLocationOnFileSystem);
+				// Overriding the value with absolute filesystem path.
+				this.configuration.put("ssl.keystore.location", fileSystemLocation);
+			}
+		}
+		catch (Exception e) {
+			throw new IllegalStateException(e);
+		}
+	}
+
+	private String moveCertToFileSystem(String classpathLocation, String fileSystemLocation) throws IOException {
+		Resource resource = new DefaultResourceLoader().getResource(classpathLocation);
+		File targetFile = new File(fileSystemLocation != null ? fileSystemLocation : "/tmp/" + resource.getFilename());
```

Comment:
I am wondering whether if this would fail when running on a Windows-based machine. Do we want to gracefully handle the scenario when `/tmp` is not an available directory?

Reply:
Good point, we should check the existence and throw a meaningful exception; we should also use `File.separator` there instead of `/`.